File: newmx/codec.py
"""
NewMx Path 1 — public Codec API.

Usage:
    from newmx import Codec

    codec = Codec()  # loads bundled v005-rev4 map by default
    encoded = codec.encode("how to set up a Docker compose file")
    print(encoded)               # → "µ set up a Docker compose file"

    full_prompt = codec.encode_with_table("how to set up a Docker compose file")
    # ↑ includes the decode-table preamble. This is what you send to the LLM.

    decoded = codec.decode(encoded)  # reverse (lossy for family glyphs)
"""

import json
import importlib.resources as resources
from dataclasses import dataclass
from typing import List, Optional, Dict

from ._normalize import normalize_line, is_code_line
from ._encoder  import compile_patterns, encode_line, CompiledPatterns
from ._decoder  import build_decode_table, build_decode_map, decode_text


@dataclass
class EncodingResult:
    """Result of encoding a single input. All fields are diagnostic; the main
    result is `.encoded` for the compressed text."""
    raw: str            # original input
    normalized: str     # post-normalization, pre-encoding
    encoded: str        # final encoded output (cli5)
    replacements: List[Dict]  # glyph substitutions made
    is_code: bool       # whether the input was code-detected (skipped encoding)


class Codec:
    """The Path 1 codec. Encodes English prompts to glyph-compressed form
    using v005-rev4 + cli5 by default.

    Parameters
    ----------
    map_path : str | None
        Filesystem path to a phrase-glyph-map JSON. If None, loads the
        bundled v005-rev4 map.
    """

    DEFAULT_MAP_RESOURCE = "path1_en_b3_v005_rev4.json"

    def __init__(self, map_path: Optional[str] = None):
        if map_path is not None:
            with open(map_path, encoding="utf-8") as f:
                self._map = json.load(f)
        else:
            # Load the bundled map from package resources
            with resources.files("newmx.maps").joinpath(
                self.DEFAULT_MAP_RESOURCE
            ).open(encoding="utf-8") as f:
                self._map = json.load(f)

        self._compiled    = compile_patterns(self._map)
        self._decode_map  = build_decode_map(self._map)
        self._decode_tbl  = None  # lazy

    @property
    def codec_version(self) -> str:
        """Version string from the loaded map's metadata."""
        return self._map.get("meta", {}).get("version", "unknown")

    @property
    def num_mappings(self) -> int:
        """Total phrase mappings in the loaded codec."""
        return len(self._map["phrase_to_family_and_glyph"])

    # ------------------------------------------------------------------
    # Encoding
    # ------------------------------------------------------------------

    def encode(self, text: str) -> str:
        """Encode text to compressed form. Returns just the encoded string.

        Code-detected lines are returned unchanged. For multiline input,
        each line is encoded independently.
        """
        if not text:
            return text
        return self._encode_full(text).encoded

    def encode_detailed(self, text: str) -> EncodingResult:
        """Encode text and return all diagnostic information."""
        return self._encode_full(text)

    def _encode_full(self, text: str) -> EncodingResult:
        # Multi-line: encode each line independently
        if "\n" in text:
            results = [self._encode_line(line) for line in text.split("\n")]
            return EncodingResult(
                raw="\n".join(r.raw for r in results),
                normalized="\n".join(r.normalized for r in results),
                encoded="\n".join(r.encoded for r in results),
                replacements=[r for result in results for r in result.replacements],
                is_code=any(r.is_code for r in results),
            )
        return self._encode_line(text)

    def _encode_line(self, line: str) -> EncodingResult:
        if is_code_line(line):
            return EncodingResult(
                raw=line, normalized=line, encoded=line,
                replacements=[], is_code=True,
            )
        normalized = normalize_line(line)
        encoded, reps = encode_line(normalized, self._compiled)
        return EncodingResult(
            raw=line, normalized=normalized, encoded=encoded,
            replacements=reps, is_code=False,
        )

    # ------------------------------------------------------------------
    # Decode-table-aware encoding (full LLM-ready prompt)
    # ------------------------------------------------------------------

    def encode_with_table(self, text: str) -> str:
        """Encode text and prepend the decode-table preamble. This is what you
        send to the LLM — the model needs the table to decode the glyphs."""
        encoded = self.encode(text)
        return f"{self.decode_table}\n\n{encoded}"

    @property
    def decode_table(self) -> str:
        """The decode-table preamble (lazy-built)."""
        if self._decode_tbl is None:
            self._decode_tbl = build_decode_table(self._map)
        return self._decode_tbl

    # ------------------------------------------------------------------
    # Decoding (reverse direction)
    # ------------------------------------------------------------------

    def decode(self, encoded: str) -> str:
        """Decode glyph-compressed text back to phrase form.

        NOTE: family glyphs decode to a canonical representative phrase, not
        the original input phrase. This is expected — Path 1 family
        compression is intentionally lossy at the surface level.
        """
        return decode_text(encoded, self._decode_map)


# Module-level convenience functions (use the bundled codec)
_default_codec: Optional[Codec] = None


def _get_default_codec() -> Codec:
    global _default_codec
    if _default_codec is None:
        _default_codec = Codec()
    return _default_codec


def encode(text: str) -> str:
    """Encode text using the bundled v005-rev4 codec. Convenience wrapper."""
    return _get_default_codec().encode(text)


def encode_with_table(text: str) -> str:
    """Encode text + prepend decode table. Convenience wrapper."""
    return _get_default_codec().encode_with_table(text)


def decode(encoded: str) -> str:
    """Decode glyph-compressed text. Convenience wrapper."""
    return _get_default_codec().decode(encoded)

File: android/app/src/main/java/com/cccstudios/newmx/codec/Codec.kt
package com.cccstudios.newmx.codec

import android.content.Context

/**
 * Public façade for the NewMx Path 1 codec.
 *
 * Mirrors Python `newmx.Codec` API:
 *   - encode(text)              → encoded text, no decode table
 *   - encodeWithTable(text)     → decode table preamble + encoded text
 *   - decode(text)              → reverse direction (lossy for families)
 *   - encodeDetailed(text)      → EncodingResult with diagnostics
 *
 * Codec instances are heavy (loads ~310KB JSON, compiles 2 large regexes).
 * Build ONCE per process and reuse via Codec.getInstance(context).
 *
 * Thread-safety: encodings are stateless; the codec can be used from any
 * thread once constructed.
 */
class Codec private constructor(
    private val map: PhraseGlyphMap,
    private val compiled: CompiledPatterns
) {
    private val decodeMap: Map<String, Pair<String, String>> by lazy {
        DecodeTable.buildDecodeMap(map)
    }

    private val cachedDecodeTable: String by lazy {
        DecodeTable.build(map)
    }

    /** Codec version string (e.g. "v005-rev4"). */
    val codecVersion: String get() = map.version

    /** Total phrase mappings in the codec (e.g. 3,135). */
    val numMappings: Int get() = map.size

    /** Decode-table preamble. Cached; built lazily on first access. */
    val decodeTable: String get() = cachedDecodeTable

    /**
     * Encode text. Multi-line input is encoded per-line. Code-detected lines
     * are returned unchanged.
     *
     * Returns just the encoded text — for the production-ready LLM prompt
     * (including decode table), call encodeWithTable().
     */
    fun encode(text: String): String {
        if (text.isEmpty()) return text
        return encodeDetailed(text).encoded
    }

    /**
     * Encode text and prepend the decode-table preamble. This is what you
     * send to the LLM.
     */
    fun encodeWithTable(text: String): String {
        return "$decodeTable\n\n${encode(text)}"
    }

    /**
     * Decode encoded text back to canonical phrase form.
     *
     * NOTE: family glyphs decode to canonical representative, not original
     * input phrase. Path 1 is intent-preserving, not byte-exact.
     */
    fun decode(encoded: String): String {
        return DecodeTable.decode(encoded, decodeMap)
    }

    /**
     * Encode with diagnostic info.
     */
    fun encodeDetailed(text: String): EncodingResult {
        if (text.isEmpty()) {
            return EncodingResult(text, text, text, emptyList(), isCode = false)
        }
        if ('\n' in text) {
            // Multi-line: encode each line independently, concat results
            val lines = text.split('\n')
            val results = lines.map { encodeOneLine(it) }
            return EncodingResult(
                raw = lines.joinToString("\n") { it },
                normalized = results.joinToString("\n") { it.normalized },
                encoded = results.joinToString("\n") { it.encoded },
                replacements = results.flatMap { it.replacements },
                isCode = results.any { it.isCode }
            )
        }
        return encodeOneLine(text)
    }

    private fun encodeOneLine(line: String): EncodingResult {
        if (Normalizer.isCodeLine(line)) {
            return EncodingResult(line, line, line, emptyList(), isCode = true)
        }
        val normalized = Normalizer.normalizeLine(line)
        val (encoded, reps) = Encoder.encodeLine(normalized, compiled).let { it.encoded to it.replacements }
        return EncodingResult(line, normalized, encoded, reps, isCode = false)
    }

    /**
     * Diagnostic result of one encode operation.
     */
    data class EncodingResult(
        val raw: String,
        val normalized: String,
        val encoded: String,
        val replacements: List<Encoder.Replacement>,
        val isCode: Boolean
    )

    companion object {
        private const val ASSET_NAME = "path1_en_b3_v005_rev4.json"

        @Volatile
        private var INSTANCE: Codec? = null

        /**
         * Get or build the codec singleton. Application context preferred to
         * avoid Activity leaks.
         *
         * First call: ~50-150ms (JSON parse + regex compile).
         * Subsequent calls: instant.
         */
        fun getInstance(context: Context): Codec {
            INSTANCE?.let { return it }
            synchronized(this) {
                INSTANCE?.let { return it }
                val built = buildFromAsset(context.applicationContext)
                INSTANCE = built
                return built
            }
        }

        /**
         * Build the codec from the bundled JSON asset. Visible for tests.
         */
        fun buildFromAsset(context: Context): Codec {
            val jsonText = context.assets.open(ASSET_NAME).bufferedReader(Charsets.UTF_8).use {
                it.readText()
            }
            return buildFromJsonString(jsonText)
        }

        /**
         * Build the codec from a JSON string. Used by unit tests with a
         * synthetic minimal map.
         */
        fun buildFromJsonString(jsonText: String): Codec {
            val map = PhraseGlyphMap.fromJson(jsonText)
            val compiled = Encoder.compilePatterns(map)
            return Codec(map, compiled)
        }
    }
}
